mtt-management/src/mtt_types.rs

```diff
@@ -0,0 +1,446 @@
+    // 5-bit sub-index (16:12) with 1-bit left shift for the 2-bits per entry
+    let sub_index = (phys_addr & 0x1_f000) >> 11;
+    (l1_index, sub_index)
+}
+
+fn get_l1_value(pfn: u64, l1_index: u64) -> u64 {
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *const u64;
+    // Safety: The 4K L1 page is allocated by the TSM, and are are using a 12-bit index.
+    unsafe { *l1_entry_addr }
+}
+
+fn update_l1_value(pfn: u64, phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let (l1_index, sub_index) = get_l1_index_and_subindex(phys_addr.bits());
+    let value = get_l1_value(pfn, l1_index);
+    // Clear the original 2-bits and OR in the new value
+    let value = (value & !(3 << sub_index)) | ((memory_type as u64) << sub_index);
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *mut u64;
+    // Safety: The 4K L1 page is allocated by the TSM, and are are using a 12-bit index.
+    unsafe {
+        *l1_entry_addr = value;
+    };
+}
+
+// The L1 is a 4KB page that's allocated by the TSM to manage the memory type (non-confidential or confidential) at a 4K-page granularity.
+// This is done by dividing the 64MB region containing the page into 16KB regions of 4K each, and the 2-bit encoding in the L1 page pointed
+// to by the L2 entry determines the type.
+// The 12-bit index into the 4K page is computed using the 14-bits [25:12] of the physical address, and then shifting right by 2
+fn set_l1_entry(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let l2_entry = get_l2_entry(phys_addr.bits());
+    match l2_entry {
+        MttL1Dir(pfn) => update_l1_value(pfn, phys_addr, memory_type),
+        _ => unreachable!(),
+    }
+}
+
+/// Changes the L2 type for 16 consecutive entries (each representing a 64MB range) for the 1GB range
+/// containing the address to the specified type.
+/// This can be called to consolidate previously fractured ranges.
+pub fn set_1gb_range_type(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let range_start = PageSize::Size1G.round_up(phys_addr.bits());
+    for i in 0u64..=15 {
+        let addr = range_start + (i * 64 * 1024 * 1024);
+        let mtt_entry = get_l2_entry(addr);
+        if matches!(mtt_entry, MttInvalidL2) {
+            panic!("Invalid MTT entry for address {addr:x}");
+        }
+        match memory_type {
+            MttMemoryType::NonConfidential => set_l2_entry(addr, MttNonConfidential1G),
+            MttMemoryType::Confidential => set_l2_entry(addr, MttConfidential1G),
+        }
+    }
+}
+
+// Changes the L2 type of each of the 16 consecutive entries for the the 1GB range containing the address
+// from 1GB to 2MBx32MB.
+// All pages in the range will retain the original MTT type (non-confidential or confidential)
+fn fracture_1gb_range(phys_addr: SupervisorPageAddr) {
+    let range_start = PageSize::Size1G.round_up(phys_addr.bits());
```

Comment:
should this fn take an L2 entry just like the fracture for L1 below?

Reply:
My understanding is that L2 entries of type 2MB simply use the bit-mask for each of the 32x2MB regions for each of the broken-up 16x64-MB regions comprising the fractured region. So, the L2 entry should be needed only when further splitting the 64MB region.